google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracerTest.java

```diff
@@ -0,0 +1,401 @@
+                        OperationType.ServerStreaming,
+                        SpanName.of("Bigtable", "ReadRows"),
+                        statsRecorderWrapper));
+
+    Lists.newArrayList(stub.readRowsCallable().call(Query.create(TABLE_ID)));
+
+    verify(statsRecorderWrapper).putGfeLatencies(longValue.capture());
+    assertThat(longValue.getValue()).isEqualTo(FAKE_SERVER_TIMING);
+
+    verify(statsRecorderWrapper, times(2)).putGfeMissingHeaders(longValue.capture());
+    assertThat(longValue.getValue()).isEqualTo(1);
+  }
+
+  @Test
+  public void testReadRowsApplicationLatencyWithAutoFlowControl() throws Exception {
+    when(mockFactory.newTracer(any(), any(), any()))
+        .thenAnswer(
+            (Answer<BuiltinMetricsTracer>)
+                invocationOnMock ->
+                    new BuiltinMetricsTracer(
+                        OperationType.ServerStreaming,
+                        SpanName.of("Bigtable", "ReadRows"),
+                        statsRecorderWrapper));
+
+    final long applicationLatency = 200;
+    final SettableApiFuture future = SettableApiFuture.create();
+    final AtomicInteger counter = new AtomicInteger(0);
+    // For auto flow control, application latency is the time application spent in onResponse.
+    stub.readRowsCallable()
+        .call(
+            Query.create(TABLE_ID),
+            new ResponseObserver<Row>() {
+              @Override
+              public void onStart(StreamController streamController) {}
+
+              @Override
+              public void onResponse(Row row) {
+                try {
+                  counter.getAndIncrement();
+                  Thread.sleep(applicationLatency);
+                } catch (InterruptedException e) {
+                }
+              }
+
+              @Override
+              public void onError(Throwable throwable) {
+                future.setException(throwable);
+              }
+
+              @Override
+              public void onComplete() {
+                future.set(null);
+              }
+            });
+    future.get();
+
+    verify(statsRecorderWrapper).putApplicationLatencies(longValue.capture());
+
+    assertThat(counter.get()).isGreaterThan(0);
+    assertThat(longValue.getValue()).isAtLeast(applicationLatency * counter.get());
+    assertThat(longValue.getValue()).isLessThan(applicationLatency * (counter.get() + 1));
+  }
+
+  @Test
+  public void testReadRowsApplicationLatencyWithManualFlowControl() throws Exception {
+    when(mockFactory.newTracer(any(), any(), any()))
+        .thenAnswer(
+            (Answer<BuiltinMetricsTracer>)
+                invocationOnMock ->
+                    new BuiltinMetricsTracer(
+                        OperationType.ServerStreaming,
+                        SpanName.of("Bigtable", "ReadRows"),
+                        statsRecorderWrapper));
+
+    final long applicationLatency = 200;
+    final AtomicInteger counter = new AtomicInteger(0);
```

Comment:
why is this atomic?

Reply:
Hmm, because it's used in response observer so it needs to be final, and I also need to update the value, so i need to  AtomicInteger here?